src/sage/rings/species.py

```diff
@@ -0,0 +1,2390 @@
+            0
+
+        We can create the table of marks for the symmetric group::
+
+            sage: E3 = P(SymmetricGroup(3))
+            sage: C = [X^3, X*E2, C3, E3]
+            sage: table([(b, [(a.hadamard_product(b)).coefficient(b.support()[0]) for a in C]) for b in C])
+              X^3     [6, 3, 2, 1]
+              X*E_2   [0, 1, 0, 1]
+              C_3     [0, 0, 2, 1]
+              E_3     [0, 0, 0, 1]
+
+        TESTS::
+
+            sage: C3.hadamard_product(-C3)
+            -2*C_3
+
+            sage: Q = PolynomialSpecies(ZZ, ["Y"])
+            sage: P.one().hadamard_product(Q.one())
+            Traceback (most recent call last):
+            ...
+            ValueError: the factors of a Hadamard product must have the same parent
+        """
+        P = self.parent()
+        if P is not other.parent():
+            raise ValueError("the factors of a Hadamard product must have the same parent")
+
+        result = P.zero()
+        # we should first collect matching multicardinalities.
+        for L, c in self:
+            mc = L.grade()
+            tc = sum(mc)
+            S = SymmetricGroup(tc).young_subgroup(mc)
+            # conjugate L and R to match S
+            G, dompart = L.permutation_group()
+            g = list(chain.from_iterable(dompart))
+            conj_L = PermutationGroupElement(g).inverse()
+            G = libgap.ConjugateGroup(G, conj_L)
+            pi = {i: range(x - mc[i] + 1, x + 1)
+                  for i, x in enumerate(accumulate(mc))}
+            for R, d in other:
+                if mc != R.grade():
+                    continue
+                G_R, dompart_R = R.permutation_group()
+                g = list(chain.from_iterable(dompart_R))
+                conj_R = PermutationGroupElement(g).inverse()
+                H = libgap.ConjugateGroup(G_R, conj_R)
+                taus = libgap.DoubleCosetRepsAndSizes(S, G, H)
+                # loop over representatives
+                new = P.zero()
+                for tau, _ in taus:
+                    G_H = libgap.Intersection(libgap.ConjugateGroup(H, tau), G)
+                    K = PermutationGroup(gap_group=G_H, domain=range(1, tc + 1))
+                    F = P(K, pi, check=False)
+                    new += F
+                result += c * d * new
+
+        return result
+
+    def _compose_with_singletons(self, names, args):
+        r"""
+        Return the inner sum of Exercise 2.6.16 in [BLL1998]_,
+        generalized to the case of arbitrary arity.
+
+        The `k`-sort species ``self`` should be homogeneous.
+
+        INPUT:
+
+        - ``names``, the (flat) list of names of the result
+        - ``args``, the sequence of `k` compositions, each of
+          which sums to the corresponding degree of ``self``,
+          where `k` is the arity of ``self``
+
+        OUTPUT:
+
+        - the polynomial species `self(X_1 + \dots + X_{m_1}, Y_1
+          + \dots + Y_{m_2}, \dots)`, where `m_i` is the number
```

Comment:
```suggestion
        - the polynomial species `self(X_1 + \cdots + X_{m_1}, Y_1
          + \cdots + Y_{m_2}, \ldots)`, where `m_i` is the number
```
Perhaps you can tell I dislike `dots`; I think it is better to be explicit.

Reply:
I like it a lot :-)